systemd/systemd_test.go

```diff
@@ -1398,6 +1398,124 @@ X-SnapdOrigin=bar
 	})
 }
 
+func (s *SystemdTestSuite) TestAddKernelModulesMountUnit(c *C) {
+	rootDir := dirs.GlobalRootDir
+
+	restore := squashfs.MockNeedsFuse(false)
+	defer restore()
+
+	mockSnapPath := filepath.Join(c.MkDir(), "/var/lib/snapd/snaps/mykernel+mykmod_11.comp")
+	makeMockFile(c, mockSnapPath)
+
+	addMountUnitOptions := &MountUnitOptions{
+		MountUnitType: KernelModulesMountUnit,
+		Lifetime:      Persistent,
+		Description:   "Mount unit for wifi kernel modules component",
+		What:          mockSnapPath,
+		Where:         "/run/mnt/kernel-modules/5.15.0-91-generic/mykmod/",
+		Fstype:        "squashfs",
+		Options:       []string{"nodev,ro,x-gdu.hide,x-gvfs-hide"},
+		Origin:        "",
+	}
+	mountUnitName, err := NewUnderRoot(rootDir, SystemMode, nil).EnsureMountUnitFileWithOptions(addMountUnitOptions)
+	c.Assert(err, IsNil)
+	defer os.Remove(mountUnitName)
+
+	c.Assert(filepath.Join(dirs.SnapServicesDir, mountUnitName), testutil.FileEquals, fmt.Sprintf(`[Unit]
+Description=Mount unit for wifi kernel modules component
+DefaultDependencies=no
+After=systemd-remount-fs.service
+Before=systemd-udevd.service systemd-modules-load.service
+Before=umount.target
+Conflicts=umount.target
+
+[Mount]
+What=%s
+Where=/run/mnt/kernel-modules/5.15.0-91-generic/mykmod/
+Type=squashfs
+Options=nodev,ro,x-gdu.hide,x-gvfs-hide
+LazyUnmount=yes
+
+[Install]
+WantedBy=sysinit.target
+`, mockSnapPath))
+	escapedUnit := "run-mnt-kernel\\x2dmodules-5.15.0\\x2d91\\x2dgeneric-mykmod.mount"
+	c.Assert(s.argses, DeepEquals, [][]string{
+		{"daemon-reload"},
+		{"--root", rootDir, "enable", escapedUnit},
+		{"reload-or-restart", escapedUnit},
+	})
+}
+
+func (s *SystemdTestSuite) TestAddKernelTreeMountUnit(c *C) {
+	rootDir := dirs.GlobalRootDir
+
+	restore := squashfs.MockNeedsFuse(false)
+	defer restore()
+
+	addMountUnitOptions := &MountUnitOptions{
+		MountUnitType: KernelTreeMountUnit,
+		Lifetime:      Persistent,
+		Description:   "Mount unit for kernel modules in kernel tree",
+		What:          "/run/mnt/kernel-modules/5.15.0-91-generic/mykmod/modules/5.15.0-91-generic",
+		Where:         "/usr/lib/modules/5.15.0-91-generic/updates/mykmod/",
+		Fstype:        "none",
+		Options:       []string{"bind"},
+		Origin:        "",
+	}
+	mountUnitName, err := NewUnderRoot(rootDir, SystemMode, nil).EnsureMountUnitFileWithOptions(addMountUnitOptions)
+	c.Assert(err, IsNil)
+	defer os.Remove(mountUnitName)
+
+	c.Assert(filepath.Join(dirs.SnapServicesDir, mountUnitName), testutil.FileEquals, fmt.Sprintf(`[Unit]
+Description=Mount unit for kernel modules in kernel tree
+DefaultDependencies=no
+After=systemd-remount-fs.service
+After=run-mnt-kernel\x2dmodules-5.15.0\x2d91\x2dgeneric-mykmod.mount
```

Comment:
isn't this implicit already even without default dependencies?

Reply:
Hm, I'd say it is not, why do you think it is implicit? The "what" is only a folder, this unit does not know otherwise that it needs to be in the component to be mounter there.

Reply:
from systemd.mount(5):
```
       •   If a mount unit is beneath another mount unit in the file system
           hierarchy, both a requirement dependency and an ordering dependency
           between both units are created automatically.
```

Reply:
Maybe I'm getting something wrong, but that would be a situation where we have, say, `dir1-dir2-dir3.mount` and `dir1-dir2.mount`, then automatic dependencies are added so the latter gets mounted before than the former. But in this case we have `run-mnt-kernel\x2components-<...>.mount` and `usr-lib-modules-<...>.mount`, where none is beneath the other. I'm not sure if systemd would calculate deps between where of one unit and a what from another, as in some cases it would be valid to change order after or before depending on what you want.

Reply:
Sorry, my bad. I was missing some context here. I missed that there's one mount unit mounting location from a snap under /run and then another one mount said /run location to one in the rootfs.

Reply:
So in the end I did some testing that showed that dependencies were being implicitly added and actually it looks like systemd does goes up not only in the "where" of the mount unit but also in the "what":
https://github.com/systemd/systemd/blob/v255/src/core/mount.c#L307-L314
, adding dependencies as it goes through parent directories.

So we can actually remove the After, which will simplify things. cc @pedronis

Reply:
Interesting, I new it dealt with the target tree dependency but wasn't aware that for bind mounts it tracks the source as well.